// rettiwt-api/dist/enums/raw/Analytics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERawAnalyticsMetric = exports.ERawAnalyticsGranularity = void 0;
/**
 * Granularity types for analytics.
 *
 * @public
 */
var ERawAnalyticsGranularity;
(function (ERawAnalyticsGranularity) {
    ERawAnalyticsGranularity["DAILY"] = "Daily";
    ERawAnalyticsGranularity["WEEKLY"] = "Weekly";
    ERawAnalyticsGranularity["MONTHLY"] = "Monthly";
})(ERawAnalyticsGranularity || (exports.ERawAnalyticsGranularity = ERawAnalyticsGranularity = {}));
/**
 * The different types of metric someone can request.
 *
 * @public
 */
var ERawAnalyticsMetric;
(function (ERawAnalyticsMetric) {
    ERawAnalyticsMetric["ENGAGEMENTS"] = "Engagements";
    ERawAnalyticsMetric["IMPRESSIONS"] = "Impressions";
    ERawAnalyticsMetric["PROFILE_VISITS"] = "ProfileVisits";
    ERawAnalyticsMetric["FOLLOWS"] = "Follows";
    ERawAnalyticsMetric["VIDEO_VIEWS"] = "VideoViews";
    ERawAnalyticsMetric["REPLIES"] = "Replies";
    ERawAnalyticsMetric["LIKES"] = "Likes";
    ERawAnalyticsMetric["RETWEETS"] = "Retweets";
    ERawAnalyticsMetric["MEDIA_VIEWS"] = "MediaViews";
    ERawAnalyticsMetric["BOOKMARK"] = "Bookmark";
    ERawAnalyticsMetric["SHARE"] = "Share";
})(ERawAnalyticsMetric || (exports.ERawAnalyticsMetric = ERawAnalyticsMetric = {}));
